### PR TITLE
[cleanup] CI test updates; mypy cleanup; partial broadcast_object cleanup; pre-commit documentation

### PR DESCRIPTION
**Pull request by min-xu-ai**
## What does this PR do?
This does a number of cleanups:

1. changelog update
2. CI test updates
3. mypy cleanup
4. broadcast_object cleanup

fixes: #737

## Before submitting

- [x] Did you have fun?
  - Make sure you had fun coding 🙃
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [x] N/A
- [ ] Did you make sure to update the docs?
  - [x] N/A
- [ ] Did you write any new necessary tests?
  - [x] N/A
- [x] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/master/CHANGELOG.md)? (if needed)
  - [ ] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

File: fairscale/optim/oss.py
Comment (min-xu-ai) [inline fairscale/optim/oss.py:288]: HI @blefaudeux, I am clean up the call of broadcast_object since we now have retired the older pytorch version that doesn't support it. However, it is causing some oss test failure. Can you help me check what might be wrong with the cleanup change?

Here is the CI failure:

I ran the failed tests locally and it worked fine. :-(

Comment (blefaudeux) [reply fairscale/optim/oss.py:288]: Oh it looks good from a quick check, could that be a backend issue (gloo/nccl) ?

Comment (blefaudeux) [reply fairscale/optim/oss.py:288]: I'm guessing that this should work for everything, not just tensors, right ?

**Comment by min-xu-ai** [comment]
This should be ready. Once the fix of #775 land, I will merge it into this branch and merge this after that fix, in addition to any review comments, of course.

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py
Comment (min-xu-ai) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1188]: this fix will come from the master branch, not this PR.

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py
Comment (min-xu-ai) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1817]: this part is updating to use standard pytorch dict.broadcast_object_list and remove mypy ignores.

File: fairscale/optim/oss.py
Comment (min-xu-ai) [inline fairscale/optim/oss.py:47]: kept this function only in oss.py for older GPUs (really only for testing).

File: CONTRIBUTING.md
Comment (blefaudeux) [inline CONTRIBUTING.md:60]: nice, thanks for this @min-xu-ai, I've been willing to write this down for a while

File: fairscale/optim/oss.py
Comment (blefaudeux) [inline fairscale/optim/oss.py:54]: oh wow, I saw that behaviour long ago but did not match it with old GPUs, well done ! :O

File: fairscale/optim/oss.py
Comment (blefaudeux) [inline fairscale/optim/oss.py:337]: personally I would be fine dumping the old gpu case (with a proper warning), I'm not sure that the intersection of people willing to shard optimizer state and the people using old GPUs is that big ?

File: tests/optim/test_oss.py
Comment (blefaudeux) [inline tests/optim/test_oss.py:45]: on the OSS side, I think that the best now would be to use the [distributed key value store](https://pytorch.org/docs/stable/distributed.html?highlight=key%20value%20store#distributed-key-value-store), in a non-blocking fashion. This could simplify the calls to .state_dict(), to my understanding, since all ranks could pull the state from the store and we could thus skip the .consolidate_state_dict(). I'm not completely sure that it's worth the work at this point in time, but just a heads up in case that's useful with FSDP

**Review by blefaudeux — APPROVED** [review]
LGTM, thanks for all this work @min-xu-ai , very valuable indeed !

File: fairscale/optim/oss.py
Comment (min-xu-ai) [reply fairscale/optim/oss.py:337]: yeah, I would be fine with that too. unfortunately, circle CI only have M60 for us for now. Maybe we will have T4 soon. Before that, we just have this local patch for now.

File: tests/optim/test_oss.py
Comment (min-xu-ai) [reply tests/optim/test_oss.py:45]: Thanks. Sounds interesting. CC @zhaojuanmao  for her info as well.

File: tests/nn/data_parallel/test_fsdp.py
Comment (zhaojuanmao) [inline tests/nn/data_parallel/test_fsdp.py:116]: I encountered the same error while upstreaming FSDP to PyTorch, I added 'ref_loss = ref_loss.cpu()' after line 105, it works

**Review by zhaojuanmao — APPROVED** [review]

File: tests/nn/data_parallel/test_fsdp.py
Comment (min-xu-ai) [reply tests/nn/data_parallel/test_fsdp.py:116]: Good to know! I just wanted to make sure we only move the device when we are expecting in the test so that it is a bit more strict.

File: fairscale/optim/oss.py
Comment (min-xu-ai) [inline fairscale/optim/oss.py:55]:
```suggestion
    the hang behavior persist across processes once it happens. I.e. once we call dist.broadcast_object_list,
```

File: CONTRIBUTING.md
Comment (min-xu-ai) [inline CONTRIBUTING.md:70]:
```suggestion
### Run statis analysis by hand (without using pre-commit)

Note that, trailing spaces are not checked by the manual commands below, but they are checked by the pre-commit hooks above.
```

**Pull request merged**
